api/dbadapters/bigquery.ts

```diff
@@ -257,54 +248,82 @@ export class BigQueryDbAdapter implements IDbAdapter {
-      })
-    );
-    const queries = actions.map(action => {
-      const persistTable = dataform.PersistedTableMetadata.create({
-        target: action.target,
-        lastUpdatedMillis: tableMetadataByTarget.get(action.target).lastUpdatedMillis,
-        definitionHash: hashExecutionAction(action),
-        transitiveInputTables: action.transitiveInputs.map(transitiveInput => {
-          if (!tableMetadataByTarget.has(transitiveInput)) {
-            throw new Error(`Could not find table metadata for ${JSON.stringify(transitiveInput)}`);
-          }
-          return tableMetadataByTarget.get(transitiveInput);
-        })
-      });
-
-      const targetName = `${action.target.database}.${action.target.schema}.${action.target.name}`;
-
-      return buildQuery(targetName, persistTable);
-    });
+    const targetStringifier = JSONObjectStringifier.create<dataform.ITarget>();
+    try {
+      // Create the cache table, if needed.
+      await this.execute(
+        `
+CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+  target STRING,
+  metadata_proto STRING
+)`,
+        options
+      );
+      // Before saving any new data, delete all entries for 'allActions'.
+      await this.execute(
+        `
+DELETE \`${CACHED_STATE_TABLE_NAME}\` WHERE target IN (${allActions
+          .map(({ target }) => `'${targetStringifier.stringify(target)}'`)
+          .join(",")})`,
+        options
+      );
 
-    const unionQuery = queries.join(" UNION ALL ");
-
-    const updateQuery = `MERGE INTO \`${CACHED_STATE_TABLE_NAME}\` T
-    USING (${unionQuery}) S
-    ON (T.target_name = S.target_name)
-    WHEN NOT MATCHED THEN
-      INSERT (target_name, metadata_json, metadata_proto)
-      VALUES(S.target_name, S.metadata_json, S.metadata_proto)
-    WHEN MATCHED THEN
-      UPDATE SET metadata_json = S.metadata_json,
-          metadata_proto = S.metadata_proto;`;
-    await this.runQuery(updateQuery);
+      // Save entries for 'actionsToPersist'.
+      const transitiveInputMetadataByTarget = await this.computeTransitiveInputMetadataMap(
+        actionsToPersist
+      );
+      const valuesTuples = actionsToPersist
+        // If we were unable to load metadata for the action's output dataset, or for any of the action's
+        // input datasets, do not store a cache entry for the action.
+        .filter(
+          action =>
+            transitiveInputMetadataByTarget.has(action.target) &&
+            action.transitiveInputs.every(transitiveInput =>
+              transitiveInputMetadataByTarget.has(transitiveInput)
+            )
+        )
+        .map(
+          action =>
+            `('${targetStringifier.stringify(action.target)}', '${encodePersistedTableMetadata({
+              target: action.target,
+              lastUpdatedMillis: transitiveInputMetadataByTarget.get(action.target)
+                .lastUpdatedMillis,
+              definitionHash: hashExecutionAction(action),
```

Comment:
A quick thought. Why do we store so much structure here, do we need it? I guess it's useful for understanding why something cached. Why not hash the entire thing - definition + (ordered) transitive inputs with last modified timestamps, then there's just one key to check?

Based on the PRD we aren't going to attempt to explain WHY an action wasn't cached. If we plan on doing that in the future though, it could be easier.

Reply:
I have considered this, but I don't think it's a good idea. If we ever have to debug why something cached after the fact, having a hashed object isn't going to help at all. I have even considered un-hashing `ExecutionAction`, but didn't do it to avoid writing potentially very large rows to the cache.